src/backend/commands/vacuum.c

```diff
@@ -2009,21 +1954,23 @@ vac_update_datfrozenxid(void)
 		 * heap table instead of system cache
 		 * "SearchSysCacheCopy1(DATABASEOID, ObjectIdGetDatum(MyDatabaseId))".
 		 * Since the cache already flatten toast tuple, so the
-		 * heap_inplace_update will fail with "wrong tuple length".
+		 * systable_inplace_update_finish will fail with "wrong tuple length".
 		 */
-		tuple = fetch_database_tuple(relation, MyDatabaseId);
+		ScanKeyInit(&key[0],
+					Anum_pg_database_oid,
+					BTEqualStrategyNumber, F_OIDEQ,
+					ObjectIdGetDatum(MyDatabaseId));
+
+		systable_inplace_update_begin(relation, DatabaseOidIndexId, true,
+									  NULL, 1, key, &tuple, &inplace_state);
+
 		if (!HeapTupleIsValid(tuple))
 			elog(ERROR, "could not find tuple for database %u", MyDatabaseId);
 		tmp_dbform = (Form_pg_database) GETSTRUCT(tuple);
 		tmp_dbform->datfrozenxid = newFrozenXid;
 		tmp_dbform->datminmxid = newMinMulti;
-=======
-		systable_inplace_update_finish(inplace_state, tuple);
-	else
-		systable_inplace_update_cancel(inplace_state);
```

Comment:
Do we need a `systable_inplace_update_cancel`?

Reply:
no, on GPDB we are in `if (dirty)` block